src/kmer_collection.h

```diff
@@ -126,57 +153,76 @@ class KmerCollection {
     m_scoreMin(std::move(other.m_scoreMin)),
     m_kmerListSize(std::move(other.m_kmerListSize)),
     m_kmerList(std::move(other.m_kmerList)),
-    m_usedGenerator(&KmerCollection::generateKmersN),
+    m_usedGenerator(&KmerCollection::generateKmerN),
     m_compress(std::move(other.m_compress))
-  {}
+  {
+    other.m_kmerList = nullptr;
+  }
 
-  virtual ~KmerCollection() {
-    if(m_kmerList) {
-      delete[] m_kmerList;
-    }
+  /**
+  * Generates kmers if necessary.
+  * \return Number of kmers.
+  */
+  size_t size() {
+    generateKmers();
+    return m_kmerListSize;
   }
 
+  /**
+  * Generates kmers if necessary.
+  * \return Iterator to KmerCollection's first element.
+  */
   KmerCollection::iterator begin() {
+    generateKmers();
+    return KmerIterator(this, 0);
+  }
+
+  /**
+  * \return Iterator to the element after KmerCollection's last element.
+  */
+  inline KmerCollection::iterator end() {
+    return KmerIterator(this, m_kmerListSize);
+  }
+
+ private:
+  inline void generateKmers() {
     if(m_kmerList == nullptr) {
       chooseGenerator();
 
-      m_kmerList = new uint_fast64_t[m_kmerListSize];
+      m_kmerListSize = (m_seqEntry->seq().size() - m_kmerSize) / m_stepSize + 1;
+
+      m_kmerList = std::unique_ptr<uint_fast64_t[]>(new uint_fast64_t[m_kmerListSize]);
       const char* ptr = m_seqEntry->seq().data();
 
-      for(int i = 0; i < m_kmerListSize; ++i) {
-        m_kmerList[i] = (this->*m_usedGenerator)(&(ptr[i]));
+      for(size_t i = 0; i < m_kmerListSize; ++i) {
+        m_kmerList[i] = (this->*m_usedGenerator)(&(ptr[i*m_stepSize]));
       }
     }
-    return KmerIterator(this, 0);
   }
 
-  KmerCollection::iterator end() {
-    return KmerIterator(this, m_kmerListSize);
-  }
-
- private:
-  void chooseGenerator() {
-    if(m_compress == false) switch(m_kmerSize) {
-      case 8: m_usedGenerator = &KmerCollection::generateUncompressedKmers8; break;
-      default: m_usedGenerator = &KmerCollection::generateUncompressedKmersN; break;
+  inline void chooseGenerator() {
+    if(!m_compress) switch(m_kmerSize) {
+      case 8: m_usedGenerator = &KmerCollection::generateUncompressedKmer8; break;
+      default: m_usedGenerator = &KmerCollection::generateUncompressedKmerN; break;
     }
     else switch(m_kmerSize) {
-      case 8: m_usedGenerator = &KmerCollection::generateKmers8; break;
-      default: m_usedGenerator = &KmerCollection::generateKmersN; break;
+      case 8: m_usedGenerator = &KmerCollection::generateKmer8; break;
+      default: m_usedGenerator = &KmerCollection::generateKmerN; break;
     }
   }
 
   // This assumes the next 8 chars is readable
-  uint_fast64_t generateKmers8(const char* ptr) {
-    uint_fast64_t kmer = *(const uint_fast64_t*)ptr;
+  uint_fast64_t generateKmer8(const char* ptr) const {
+    const uint_fast64_t mask = 0x0606060606060606;
+    uint64_t kmer = *(const uint64_t*)ptr & mask;
```

Comment:
This part is still extremely hacky, but i do not see a way to make it less hacky and still keep the speed.